src/Common/ZooKeeper/ZooKeeperImpl.h

```diff
@@ -239,8 +239,13 @@ class ZooKeeper final : public IKeeper
     Poco::Net::StreamSocket socket;
     /// To avoid excessive getpeername(2) calls.
     Poco::Net::SocketAddress socket_address;
-    std::optional<ReadBufferFromPocoSocket> in;
-    std::optional<WriteBufferFromPocoSocket> out;
+
+    std::shared_ptr<ReadBufferFromPocoSocket> in;
+    std::shared_ptr<WriteBufferFromPocoSocket> out;
+    std::shared_ptr<ReadBuffer> maybe_compressed_in;
+    std::shared_ptr<WriteBuffer> maybe_compressed_out;
```

Comment:
Actually only `maybe_compressed_out->next()`; should be enough (as it will have the normal write buffer if not compressed).  I will remove `out->next()`

Reply:
Turns out it is needed. I will update as you suggested add use `use_compression` for the check